extensions/2.0/Khronos/KHR_materials_specular/README.md

```diff
@@ -0,0 +1,177 @@
+
+```json
+{
+    "materials": [
+        {
+            "extensions": {
+                "KHR_materials_specular": {
+                    "specularFactor": 1.0,
+                    "specularColorFactor": [1.0, 1.0, 1.0],
+                    "specularTexture": 0,
+                }
+            }
+        }
+    ]
+}
+```
+
+Factor and texture are combined by multiplication to describe a single value.
+
+| |Type|Description|Required|
+|-|----|-----------|--------|
+| **specularFactor** | `number` | The strength of the specular reflection. | No, default: `1.0`|
+| **specularTexture** | [`textureInfo`](/specification/2.0/README.md#reference-textureInfo) | A grayscale texture that defines the specular factor. Will be multiplied by specularFactor. | No |
+| **specularColorFactor** | `number[3]` | The F0 color of the specular reflection (RGB). | No, default: `[1.0, 1.0, 1.0]`|
+| **specularColorTexture** | [`textureInfo`](/specification/2.0/README.md#reference-textureInfo) | A texture that defines the F0 color of the specular reflection (RGB channels, encoded in sRGB) and the specular factor (A). Will be multiplied by specularFactor and specularColorFactor. | No |
+
+The specular factor scales the microfacet BRDF in the dielectric BRDF. It also affects the diffuse BRDF; the less energy is reflected by the microfacet BRDF, the more can be shifted to the diffuse BRDF. The following image shows specular factor increasing from 0 to 1.
+
+![](figures/specular.png)
+
+The specular color changes the F0 color of the Fresnel that is multiplied to the microfacet BRDF. The color at grazing angles (F90) is not changed. As with specular factor, the diffuse BRDF will take the remaining energy, but in this case it is directional-dependent. The following images show specular color increasing from [0,0,0] to [1,1,1] (top) and from [0,0,0] to [1,0,0] (bottom).
+
+![](figures/specular-color.png)
+![](figures/specular-color-2.png)
+
+## Implementation
+
+*This section is non-normative.*
+
+The extension changes the computation of the Fresnel term defined in [Appendix B](/specification/2.0/README.md#appendix-b-brdf-implementation) to the following:
+
+```
+dielectricSpecularF0 = 0.04 * specularFactor * specularTexture.a *
+                              specularColorFactor * specularColorTexture.rgb
+dielectricSpecularF90 = specularFactor * specularTexture.a
+
+F0  = lerp(dielectricSpecularF0, baseColor.rgb, metallic)
+F90 = lerp(dielectricSpecularF90, 1, metallic)
+
+F = F0 + (F90 - F0) * (1 - VdotH)^5
+```
+
+## Interaction with other extensions
+
+If `KHR_materials_ior` is used in combination with `KHR_materials_specular`, the constant `0.04` is replaced by the value computed from the IOR.
+
+```
+dielectricSpecularF0 = ((ior - outside_ior) / (ior + outside_ior))^2 * specularFactor * specularTexture.a * specularColorFactor * specularColorTexture.rgb
```

Comment:
It seems the specularTexture.rgb could be pre-multiplied with specularTexture.a and the specularTexture.a contain the premultiplied specularFactor * specularTexture.a - or am I missing something?

Reply:
It is indeed possible to premultiply specular and specularColor and store that in specularTexture.rgb. It saves one multiply instruction. But I would not recommend it.

With premultiplication, the values in the texture will have a different semantic than the scalar parameters. This makes it harder to understand and increases complexity in tools. In addition, it is not guaranteed anymore that `dielectricSpecularF0 < dielectricSpecularF90`, a property that would then have to be enforced in the shader, requiring additional instructions somewhere else.